File: bindings/browser/examples/encode-to-mp4.ts
// encode-to-mp4.ts — Mediaway browser (WASM) quick start (ASPIRATIONAL EXAMPLE).
//
// No `@mediaway/browser` package exists yet — this file shows the target
// ergonomics for a future browser binding wrapping Mediaway's WASM module
// (`wasm-bindgen`), never the C ABI. See ../README.md and docs/spec/c-ffi.md
// Tier C.
//
// Mirrors examples/encode_to_mp4.rs: build a 640x480 @30fps H.264 config,
// open the best available auto encode path in this browser, push 90
// synthetic grey NV12 frames (3 s at 30 fps) through an encode session, and
// get back the resulting fragmented MP4 bytes.
//
// Browser-specific note: real H.264 hardware encode on the web goes through
// the browser's native `WebCodecs` `VideoEncoder`, not a software encoder
// reimplemented in WASM. `AutoVideoEncoder` is expected to drive WebCodecs
// under the hood and hand the resulting packets to Mediaway's WASM
// fragmented-MP4 muxer — callers still see the same "config -> open encoder
// -> encode session -> writeFrame -> finish" shape as every other language
// binding.

import init, { AutoVideoEncoder, EncodeSession } from "@mediaway/browser";
import type { VideoEncodeConfig, VideoFrame } from "@mediaway/browser";

const WIDTH = 640;
const HEIGHT = 480;
const FPS = 30;
const SECONDS = 3;
const FRAME_COUNT = FPS * SECONDS; // 90 frames = 3 s at 30 fps

async function main(): Promise<void> {
  // The WASM module must be instantiated (fetched + compiled) once before any
  // other call. Every call after this resolves is synchronous-feeling.
  await init();

  // Defaults for H264 at this resolution/framerate, then override bitrate.
  const config: VideoEncodeConfig = {
    codec: "h264",
    width: WIDTH,
    height: HEIGHT,
    frameRate: { num: 1, den: FPS },
    bitrateBps: 2_000_000,
  };

  // 1. Open the auto encoder — tries the best available encode path in this
  // browser (WebCodecs hardware/software H.264 under the hood) and throws
  // if none is available. This is a normal, expected outcome on
  // unsupported browsers, not a crash — catch it and bail out gracefully.
  let encoder: AutoVideoEncoder;
  try {
    encoder = await AutoVideoEncoder.open(config);
  } catch (err) {
    console.error(`encode-to-mp4: no supported H.264 encode path in this browser (${err})`);
    return;
  }

  console.log("encode-to-mp4: running in this browser");

  try {
    // 2. Wrap the opened encoder in an encode session — wires WebCodecs
    // output packets into the internal fragmented-MP4 muxer.
    const session = new EncodeSession(encoder);

    try {
      // Synthetic NV12 source (replace with real frames in your app): grey
      // Y=128 everywhere, U/V=128 everywhere. Layout is width*height Y bytes
      // followed by width*height/2 interleaved UV bytes.
      const ySize = WIDTH * HEIGHT;
      const uvSize = (WIDTH * HEIGHT) / 2;
      const nv12Frame = new Uint8Array(ySize + uvSize).fill(128);

      for (let pts = 0; pts < FRAME_COUNT; pts++) {
        const frame: VideoFrame = {
          pts,
          duration: 1,
          width: WIDTH,
          height: HEIGHT,
          pixelFormat: "nv12",
          data: nv12Frame,
        };
        await session.writeFrame(frame);
      }

      // 3. Flush the encoder, finalize the muxer, and get the complete MP4
      // file bytes back.
      const mp4Bytes = await session.finish();

      console.log(`encode-to-mp4: ${FRAME_COUNT} frames -> ${mp4Bytes.byteLength} bytes of fMP4`);

      // 4. Session/muxer never touch storage themselves — the caller decides
      // where the bytes go. Here: trigger a normal browser download.
      downloadAsFile(mp4Bytes, "out.mp4");
    } finally {
      session.free(); // release the WASM-side session; JS GC can't reach into WASM memory
    }
  } finally {
    encoder.free(); // release the WASM-side encoder handle
  }
}

// Trigger a browser download of the muxed bytes via an object URL + a
// synthetic anchor click.
function downloadAsFile(bytes: Uint8Array, filename: string): void {
  const blob = new Blob([bytes], { type: "video/mp4" });
  const url = URL.createObjectURL(blob);
  const anchor = document.createElement("a");
  anchor.href = url;
  anchor.download = filename;
  anchor.click();
  URL.revokeObjectURL(url);
}

main().catch((err: unknown) => {
  console.error(err);
});

File: examples/pipeline/encode_to_mp4.rs
//! High-level encode pipeline: frames → H.264 → fragmented MP4.
//!
//! No `#[cfg(…)]` here — platform selection (`mediaway_pipeline::platform`) and the
//! encoder→muxer wiring (`mediaway_pipeline::EncodeSession`) are both handled by
//! `mediaway-pipeline`; this file is just the per-frame loop.
//!
//! For the low-level path (manual poll loops, no convenience wrapper), see
//! `mux_roundtrip.rs`.
//!
//! Run:
//! ```text
//! cargo run --example encode_to_mp4
//! ```
//! Output: `out.mp4` in the current directory.

#![allow(
    clippy::print_stdout,
    clippy::print_stderr,
    clippy::expect_used,
    reason = "example demonstrates the happy path with console output"
)]

use mediaway_common::{Bytes, CodecKind, PixelFormat, Rational, VideoFrame, VideoFrameStorage};
use mediaway_encoder::auto::AutoVideoEncodeConfig;
use mediaway_pipeline::{EncodeSession, platform};
use std::fs::File;
use std::io::Write as _;

const WIDTH: u32 = 640;
const HEIGHT: u32 = 480;
const FPS: u32 = 30;
const SECONDS: u32 = 3;

fn main() {
    let config = AutoVideoEncodeConfig {
        bitrate_bps: 2_000_000,
        ..AutoVideoEncodeConfig::new(CodecKind::H264, WIDTH, HEIGHT, Rational::new(1, FPS))
    };

    let encoder = match platform::AutoEncoder::open(&config) {
        Ok(e) => e,
        Err(e) => {
            eprintln!("encode_to_mp4: open failed ({e}) — platform not supported yet");
            return;
        }
    };

    println!("encode_to_mp4: running on this platform");
    let mut session = EncodeSession::open(encoder).expect("open encode session");

    // ── Synthetic NV12 source (replace with real frames in your app) ─────────
    let nv12_len = (WIDTH * HEIGHT + WIDTH * HEIGHT / 2) as usize;
    let source = Bytes::from(vec![128u8; nv12_len]); // grey Y=128, UV=128

    for pts in 0..i64::from(FPS * SECONDS) {
        let frame = VideoFrame {
            pts,
            duration: 1,
            width: WIDTH,
            height: HEIGHT,
            format: PixelFormat::Nv12,
            storage: VideoFrameStorage::Cpu {
                // clone: Bytes ref-count bump — backing buffer is not copied
                data: source.clone(),
            },
        };
        session.write_frame(&frame).expect("write frame");
    }

    let mp4_bytes = session.finish().expect("finish encode session");

    File::create("out.mp4")
        .and_then(|mut f| f.write_all(&mp4_bytes))
        .expect("write out.mp4");

    println!(
        "encode_to_mp4: {} frames → out.mp4 ({} bytes)",
        FPS * SECONDS,
        mp4_bytes.len()
    );
}
